### PR TITLE
Add sfu example with websocket

### LINKED ISSUE #507: SFU: new sfu example with websocket
I'm writing sfu example with websocket, support video and audio. easy to use.

### PR DESCRIPTION
**Pull request by adwpc**
sfu example with websocket, support video and audio. easy to use

Resolves #507

### CODE REVIEW

**Comment by Sean-Der** [comment]
Fantastic work @adwpc 

Reviewing now :)

File: examples/sfu-ws/main.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	}()
+
+	// Websocket handle func
+	http.HandleFunc("/ws", Room)
+
+	// Html handle func
+	http.HandleFunc("/", web)
+
+	// Support https, so we can test by lan
+	if err := http.ListenAndServeTLS(":443", "cert.pem", "key.pem", nil); err != nil {
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Is it possible to generate the cert/key in the example? I don't want to encourage the reuse (or committing secrets in git)

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,7 @@
 module github.com/pions/webrtc
 
 require (
+	github.com/gorilla/websocket v1.4.0
```
Comment (Sean-Der) [inline go.mod]: can you put this dependency in the examples directory, I don't want to make this mandatory to users who don't want this library.

**Comment by Sean-Der** [comment]
Would you be interested in starting a `pions/sfu` project @adwpc ?

I personally really want to build an SFU, but haven't had time to start. I think it would be a super popular project, and I want to solve problems like auto-scaling that no other project is tackling.

File: examples/sfu-ws/main.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+		// Sets the LocalDescription, and starts our UDP listeners
+		checkError(PubReceiver.SetLocalDescription(answer))
+
+		// Send server sdp to publisher
+		checkError(c.WriteMessage(mt, []byte(answer.SDP)))
+		fmt.Println("send pion sdp to publisher:==============\n" + answer.SDP)
+	} else {
+		fmt.Println("scriber sdp:=============\n", string(msg))
+
+		recvOnlyOffer := webrtc.SessionDescription{
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Can you inline move this variable by where it is used?

File: examples/sfu-ws/main.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	},
+}
+
+var (
+	// API object
+	api *webrtc.API
+
+	// Publisher Peer
+	hasPublisher bool
+	PubReceiver  *webrtc.PeerConnection
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Does this need to be exported outside `onTrack`?

File: examples/sfu-ws/main.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+	// Local track
+	videoTrack *webrtc.Track
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Not thread safe

File: examples/sfu-ws/main.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+	audioTrack *webrtc.Track
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Not thread safe

File: examples/sfu-ws/main.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+		},
+	},
+}
+
+var (
+	// API object
+	api *webrtc.API
+
+	// Publisher Peer
+	hasPublisher bool
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: Not thread safe

File: examples/sfu-ws/main.go (hunk 1 above)
Comment (Sean-Der) [reply examples/sfu-ws/main.go]: Can you just do `panic(http.ListenAndServerTLS())` that function should never return (unless we have an error)

File: examples/sfu-ws/main.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+func Room(w http.ResponseWriter, r *http.Request) {
+
+	// Websocket client
+	c, err := upgrader.Upgrade(w, r, nil)
+	checkError(err)
+	defer c.Close()
+
+	// Read sdp from websocket
+	mt, msg, err := c.ReadMessage()
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: This `err` needs to be checked I think

File: examples/sfu-ws/main.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+	if r.Method == "GET" {
+		t, _ := template.ParseFiles("sfu.html")
+		checkError(t.Execute(w, nil))
+	}
+}
+
+func main() {
+
+	// Enable pprof
+	go func() {
```
Comment (Sean-Der) [inline examples/sfu-ws/main.go]: I would remove `pprof` support for now. I would love this documented though! So we can help users in the future if they run into issues.

File: examples/sfu-ws/main.go (hunk 1 above)
Comment (Sean-Der) [reply examples/sfu-ws/main.go]: Can you make the port configurable, I never run others code as `root` just to be safe.

File: examples/sfu-ws/main.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+	} else {
+		fmt.Println("scriber sdp:=============\n", string(msg))
+
+		recvOnlyOffer := webrtc.SessionDescription{
+			SDP:  string(msg),
+			Type: webrtc.SDPTypeOffer,
+		}
+
+		// Create a new PeerConnection
+		SubSender, err := api.NewPeerConnection(peerConnectionConfig)
```
Comment (maxhawkins) [inline examples/sfu-ws/main.go]: change `SubSender` to `subSender` to indicate it's a local variable

File: examples/sfu-ws/main.go (hunk 1 above)
Comment (adwpc) [reply examples/sfu-ws/main.go]: Good idea, i will fix.

File: go.mod (hunk 2 above)
Comment (adwpc) [reply go.mod]: Good idea, i will fix.

File: examples/sfu-ws/main.go (hunk 3 above)
Comment (adwpc) [reply examples/sfu-ws/main.go]: Yes, i'll fix

**Comment by adwpc** [comment]
> Would you be interested in starting a `pions/sfu` project @adwpc ?
> 
> I personally really want to build an SFU, but haven't had time to start. I think it would be a super popular project, and I want to solve problems like auto-scaling that no other project is tackling.

Yes, there will be many modules

**Comment by Sean-Der** [comment]
Merged! https://github.com/pions/webrtc/commit/fb72f20f663e1f61340232c34d33d5c386778866

Nice work @adwpc this is awesome, I think this will help a lot of people :)

**Pull request closed without merging**